pkg/checks/traceroute/traceroute.go

```diff
@@ -2,197 +2,263 @@ package traceroute
+// readIcmpMessage reads a packet from the provided icmp Connection. If the packet is 'Time Exceeded',
+// it reads the address of the router that dropped created the icmp packet. It also reads the source port
+// from the payload and finds the source port used by the previous tcp connection. If any error is returned,
+// an icmp packet was either not received, or the received packet was not a time exceeded.
+func readIcmpMessage(icmpListener *icmp.PacketConn, timeout time.Duration) (int, net.Addr, error) {
+	// Expected to fail due to TTL expiry, listen for ICMP response
+	if err := icmpListener.SetReadDeadline(time.Now().Add(timeout)); err != nil {
+		return 0, nil, fmt.Errorf("failed to set icmp read deadline: %w", err)
+	}
+	buffer := make([]byte, 1500) //nolint:mnd // Standard MTU size
+	n, routerAddr, err := icmpListener.ReadFrom(buffer)
+	if err != nil {
+		// we probably timed out so return
+		return 0, nil, fmt.Errorf("failed to read from icmp connection: %w", err)
 	}
-}
 
-type Traceroute struct {
-	checks.CheckBase
-	config     Config
-	traceroute tracerouteFactory
-}
+	// Parse the ICMP message
+	msg, err := icmp.ParseMessage(ipv4.ICMPTypeTimeExceeded.Protocol(), buffer[:n])
+	if err != nil {
+		return 0, nil, err
+	}
 
-type tracerouteFactory func(dest string, port, timeout, retries, maxHops int) (traceroute.TracerouteResult, error)
+	// Ensure the message is an ICMP Time Exceeded message
+	if msg.Type != ipv4.ICMPTypeTimeExceeded {
+		return 0, nil, errors.New("message is not 'Time Exceeded'")
+	}
 
-func newTraceroute(dest string, port, timeout, retries, maxHops int) (traceroute.TracerouteResult, error) {
-	opts := &traceroute.TracerouteOptions{}
-	opts.SetTimeoutMs(timeout)
-	opts.SetRetries(retries)
-	opts.SetMaxHops(maxHops)
-	opts.SetPort(port)
-	return traceroute.Traceroute(dest, opts)
-}
+	// The first 20 bytes of Data are the IP header, so the TCP segment starts at byte 20
+	tcpSegment := msg.Body.(*icmp.TimeExceeded).Data[20:]
 
-type result struct {
-	// The minimum number of hops required to reach the target
-	NumHops int
-	// The path taken to the destination
-	Hops []hop
-}
+	// Extract the source port from the TCP segment
+	destPort := int(tcpSegment[0])<<8 + int(tcpSegment[1])
 
-type hop struct {
-	Addr    string
-	Latency time.Duration
-	Success bool
+	return destPort, routerAddr, nil
 }
 
-// Run runs the check in a loop sending results to the provided channel
-func (tr *Traceroute) Run(ctx context.Context, cResult chan checks.ResultDTO) error {
-	ctx, cancel := logger.NewContextWithLogger(ctx)
-	defer cancel()
-	log := logger.FromContext(ctx)
+// TraceRoute performs a traceroute to the specified host using TCP and listens for ICMP Time Exceeded messages using ICMP.
+func TraceRoute(ctx context.Context, cfg tracerouteConfig) (map[int][]Hop, error) {
+	// this could also be a 2d array, but I feel like using an int map here makes the json easier to understand
+	// as it explicitly shows a mapping of ttl->hops
```

Comment:
fine... pls change the comment here sth like `mapping of a specific hop to its ttl (map key)`